src/library_html5.js

```diff
@@ -3269,4 +3237,56 @@ var LibraryJSEvents = {
   }
 };
 
+function handleWebGLProxying(funcs) {
+  if (!USE_PTHREADS) return; // No proxying needed in singlethreaded builds
+
+  function listOfNFunctionArgs(func) {
+    var args = [];
+    for(var i = 0; i < func.length; ++i) {
+      args.push('p' + i);
+    }
+    return args;
+  }
+
+  for(var i in funcs) {
+    if (funcs[i + '__proxy'] == 'sync_on_webgl_context_handle_thread') {
+#if OFFSCREENCANVAS_SUPPORT
+      // With OffscreenCanvas builds, GL context handle may be owned by main thread, the calling pthread,
+      // or another pthread. Route the call to the right thread.
+      // TODO: this handles the calling pthread and main thread cases, but not yet the case from pthread->pthread.
+      funcs[i + '_calling_thread'] = funcs[i];
+      funcs[i + '_main_thread'] = i + '_calling_thread';
+      funcs[i + '_main_thread__proxy'] = 'sync';
+      funcs[i + '_main_thread__sig'] = funcs[i + '__sig'];
+      if (!funcs[i + '__deps']) funcs[i + '__deps'] = [];
+      funcs[i + '__deps'].push(i + '_calling_thread');
+      funcs[i + '__deps'].push(i + '_main_thread');
+      delete funcs[i + '__proxy'];
+      var funcArgs = listOfNFunctionArgs(funcs[i]);
+      var funcArgsString = funcArgs.join(',');
+      var funcBody = `return GL.contexts[p0] ? _${i}_calling_thread(${funcArgsString}) : _${i}_main_thread(${funcArgsString});`;
```

Comment:
the comment earlier says this may need to run on the same thread, the main thread, or another pthread. How is the case of another pthread handled? IIUC this line handles the other two.

Reply:
That is the TODO in this block. It is an existing limitation that has never worked. For a few functions (canvas size setting and getting) I have manually written the pthread->pthread proxying path, it is considerably more complex.

Reply:
I see, sgtm then.